# Methods of data preprocessing in machine learning
## Lab 

#6 Working with a data set and its application for training an artificial neural network

## Objective: To train an artificial neural network using a dataset containing quantitative and categorical data.

### Work plan.

1. Open a terminal in the interactive development environment and install the sklearn (Scikit-learn) library using the pip install command.
2. In the editor of the interactive development environment (it is recommended to work in Jupyter Notebook or Spyder), create a new file (Python 3 file). In this file, using the Pandas library tools, load the dataset "pd_speech_features.csv".
3. Collect information on the dataset using the commands given in the previous labs.
4. Check the dataset for missing, empty values ​​(the command "isnull().sum()". See which attributes have the correct data type, and which ones need to be changed.
5. After processing the dataset, you need to define two variables. Save the attributes from the dataset (features) in the variable "x", and the values ​​from the target column "class" in the variable "y".
6. Split the data from "x" and "y" into a training and test sample. To do this, use the function "train_test_split". It takes the two previously specified variables x and y as arguments, and here we also determine the size of the test sample (usually no more than 20%). To do this, we assign our value in the format "0.20" (which means 20%) to the named argument "test_size". In the parameter "random_state" we set a random seed for the random number generator for the sake of consistency.
7. After preparing the dataset, design a neural network model. To do this, create a class named "Perceptron". For this class, define the "build" method with two parameters: 1) dim (input size of the training sample), 2) num_classes - the number of classes (in our case, there are 2). Next, in the body of this method, create a variable model, where we save an instance of the Sequential class from the Keras-Tensorflow library. Then, using the "add" method, we build the fully connected layers (Dense) of the neural network one by one. The first layer (input layer) will have 128 nodes (neurons) and the ReLU activation function. This layer also accepts the above-mentioned dim argument as a parameter. The second (hidden) layer has 32 neurons and the ReLU activation function. The third layer that completes the network has 2 neurons (according to the number of classes) and the Softmax activation function. 8. Now create an instance of the "Perceptron" class, saving it in a variable named "model" (you can take any name for the variable, but it is recommended to specify names that will most clearly describe their content). Don't forget to define the parameters of the "Perceptron" class: num_classes should match the number of classes in the dataset, and for dim we pass the input sample size "x_train.shape[1]", which is 753.
9. Compile the newly created neural network model. To do this, use the "compile" method (Keras-Tensorflow). As its first parameter, specify which error function to use (for example, "sparse_categorical_crossentropy", which calculates the crossentropy loss between labels and predictions). As the second parameter, specify the optimizer (for example, "adam" or stochastic gradient descent). For information on loss functions and optimizers, see the Keras, Tensorflow documentation. And as the third parameter, we specify the target metric - accuracy.
10. Now, after completing all the preparatory operations, you can begin the process of training the ANN model. To do this, apply the fit method (Keras, Tensorflow) to the model variable and specify the samples to be used (x_train, y_train), batch size, number of iterations (epochs), size of the validation sample "validation_split" (which is created automatically by this fit method) as parameters, and also define the verbose parameter (by setting it to 0, 1 or 2, we choose how we want to "see" the training progress for each epoch).
11. After completing the training process, evaluate its quality. To do this, apply the "evaluate" method (Keras, Tensorflow) to model.
12. Use the matplotlib library to plot the training and testing graphs for the model.
13. Try to standardize the training and testing samples. To do this, apply the StandardScaler function from the sklearn library. For comparison, apply the standardized sample to the ANN model, retrain it and plot graphs, evaluate the quality metrics, compare them with the previous results, whether the accuracy has improved after standardizing the data set.
14. Try training the model with other parameters: change the error function, optimizer (choose from those presented in the Keras documentation), the number of epochs (iterations). Also try changing the number of neurons in the input and hidden layers. Analyze and compare the results, plot graphs.
15. Make a final conclusion - with which parameters the best metrics for the quality of ANN training were obtained and whether standardizing the data set helped.
16. Generate a report with screenshots and the necessary explanations.

## Progress of work.

## 1. Open a terminal in the interactive development environment and install the sklearn (Scikit-learn) library using the pip install command.

In [13]:
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.model_selection import cross_val_score

## 2. In the editor of the interactive development environment (it is recommended to work in Jupyter Notebook or Spyder), create a new file (Python 3 file). In this file, using the Pandas library tools, load the data set "pd_speech_features.csv".

In [16]:
import pandas as pd
import numpy as np

df = pd.read_csv('/home/nikita/Рабочий стол/MTUCI/мл/labs1/lab6/НУ_ЛР_6(паркинсон_речь).csv', sep=',', na_values=None,header=0) #? – неизвестно;
#print (df[0:60:10])

In [18]:
df.head()

id  gender      PPE      DFA     RPDE  numPulses  numPeriodsPulses  \
0   0       1  0.85247  0.71826  0.57227        240               239   
1   0       1  0.76686  0.69481  0.53966        234               233   
2   0       1  0.85083  0.67604  0.58982        232               231   
3   1       0  0.41121  0.79672  0.59257        178               177   
4   1       0  0.32790  0.79782  0.53028        236               235   

   meanPeriodPulses  stdDevPeriodPulses  locPctJitter  ...  \
0          0.008064            0.000087       0.00218  ...   
1          0.008258            0.000073       0.00195  ...   
2          0.008340            0.000060       0.00176  ...   
3          0.010858            0.000183       0.00419  ...   
4          0.008162            0.002669       0.00535  ...   

   tqwt_kurtosisValue_dec_28  tqwt_kurtosisValue_dec_29  \
0                     1.5620                     2.6445   
1                     1.5589                     3.6107   
2                     1.5643                     2.3308   
3                     3.7805                     3.5664   
4                     6.1727                     5.8416   

   tqwt_kurtosisValue_dec_30  tqwt_kurtosisValue_dec_31  \
0                     3.8686                     4.2105   
1                    23.5155                    14.1962   
2                     9.4959                    10.7458   
3                     5.2558                    14.0403   
4                     6.0805                     5.7621   

   tqwt_kurtosisValue_dec_32  tqwt_kurtosisValue_dec_33  \
0                     5.1221                     4.4625   
1                    11.0261                     9.5082   
2                    11.0177                     4.8066   
3                     4.2235                     4.6857   
4                     7.7817                    11.6891   

   tqwt_kurtosisValue_dec_34  tqwt_kurtosisValue_dec_35  \
0                     2.6202                     3.0004   
1                     6.5245                     6.3431   
2                     2.9199                     3.1495   
3                     4.8460                     6.2650   
4                     8.2103                     5.0559   

   tqwt_kurtosisValue_dec_36  class  
0                    18.9405      1  
1                    45.1780      1  
2                     4.7666      1  
3                     4.0603      1  
4                     6.1164      1  

[5 rows x 755 columns]

In [20]:
df.tail()

id  gender      PPE      DFA     RPDE  numPulses  numPeriodsPulses  \
751  250       0  0.80903  0.56355  0.28385        417               416   
752  250       0  0.16084  0.56499  0.59194        415               413   
753  251       0  0.88389  0.72335  0.46815        381               380   
754  251       0  0.83782  0.74890  0.49823        340               339   
755  251       0  0.81304  0.76471  0.46374        340               339   

     meanPeriodPulses  stdDevPeriodPulses  locPctJitter  ...  \
751          0.004627            0.000052       0.00064  ...   
752          0.004550            0.000220       0.00143  ...   
753          0.005069            0.000103       0.00076  ...   
754          0.005679            0.000055       0.00092  ...   
755          0.005676            0.000037       0.00078  ...   

     tqwt_kurtosisValue_dec_28  tqwt_kurtosisValue_dec_29  \
751                     3.0706                     3.0190   
752                     1.9704                     1.7451   
753                    51.5607                    44.4641   
754                    19.1607                    12.8312   
755                    62.9927                    21.8152   

     tqwt_kurtosisValue_dec_30  tqwt_kurtosisValue_dec_31  \
751                     3.1212                     2.4921   
752                     1.8277                     2.4976   
753                    26.1586                     6.3076   
754                     8.9434                     2.2044   
755                     9.2457                     4.8555   

     tqwt_kurtosisValue_dec_32  tqwt_kurtosisValue_dec_33  \
751                     3.5844                     3.5400   
752                     5.2981                     4.2616   
753                     2.8601                     2.5361   
754                     1.9496                     1.9664   
755                     3.0551                     3.0415   

     tqwt_kurtosisValue_dec_34  tqwt_kurtosisValue_dec_35  \
751                     3.3805                     3.2003   
752                     6.3042                    10.9058   
753                     3.5377                     3.3545   
754                     2.6801                     2.8332   
755                     4.0116                     2.6217   

     tqwt_kurtosisValue_dec_36  class  
751                     6.8671      0  
752                    28.4170      0  
753                     5.0424      0  
754                     3.7131      0  
755                     3.1527      0  

[5 rows x 755 columns]

## 3. Collect information on the data set by using the commands given in the previous labs.

In [23]:
print(type(df))

<class 'pandas.core.frame.DataFrame'>


In [25]:
df.shape

(756, 755)

In [27]:
df.isnull().sum()

id                           0
gender                       0
PPE                          0
DFA                          0
RPDE                         0
                            ..
tqwt_kurtosisValue_dec_33    0
tqwt_kurtosisValue_dec_34    0
tqwt_kurtosisValue_dec_35    0
tqwt_kurtosisValue_dec_36    0
class                        0
Length: 755, dtype: int64

In [29]:
df_describe = df.describe(include='all')
df_describe.head(10)

id      gender         PPE         DFA        RPDE   numPulses  \
count  756.000000  756.000000  756.000000  756.000000  756.000000  756.000000   
mean   125.500000    0.515873    0.746284    0.700414    0.489058  323.972222   
std     72.793721    0.500079    0.169294    0.069718    0.137442   99.219059   
min      0.000000    0.000000    0.041551    0.543500    0.154300    2.000000   
25%     62.750000    0.000000    0.762833    0.647053    0.386537  251.000000   
50%    125.500000    1.000000    0.809655    0.700525    0.484355  317.000000   
75%    188.250000    1.000000    0.834315    0.754985    0.586515  384.250000   
max    251.000000    1.000000    0.907660    0.852640    0.871230  907.000000   

       numPeriodsPulses  meanPeriodPulses  stdDevPeriodPulses  locPctJitter  \
count        756.000000        756.000000          756.000000    756.000000   
mean         322.678571          0.006360            0.000383      0.002324   
std           99.402499          0.001826            0.000728      0.002628   
min            1.000000          0.002107            0.000011      0.000210   
25%          250.000000          0.005003            0.000049      0.000970   
50%          316.000000          0.006048            0.000077      0.001495   
75%          383.250000          0.007528            0.000171      0.002520   
max          905.000000          0.012966            0.003483      0.027750   

       ...  tqwt_kurtosisValue_dec_28  tqwt_kurtosisValue_dec_29  \
count  ...                 756.000000                 756.000000   
mean   ...                  26.237251                  22.840337   
std    ...                  42.220693                  32.626464   
min    ...                   1.509800                   1.531700   
25%    ...                   2.408675                   3.452800   
50%    ...                   5.586300                   7.062750   
75%    ...                  28.958075                  29.830850   
max    ...                 239.788800                 203.311300   

       tqwt_kurtosisValue_dec_30  tqwt_kurtosisValue_dec_31  \
count                 756.000000                 756.000000   
mean                   18.587888                  13.872018   
std                    25.537464                  20.046029   
min                     1.582900                   1.747200   
25%                     3.354825                   3.077450   
50%                     6.077400                   4.770850   
75%                    21.944050                  13.188000   
max                   121.542900                 102.207000   

       tqwt_kurtosisValue_dec_32  tqwt_kurtosisValue_dec_33  \
count                 756.000000                 756.000000   
mean                   12.218953                  12.375335   
std                    17.783642                  16.341665   
min                     1.789500                   1.628700   
25%                     2.937025                   3.114375   
50%                     4.300450                   4.741450   
75%                    10.876150                  12.201325   
max                    85.571700                  73.532200   

       tqwt_kurtosisValue_dec_34  tqwt_kurtosisValue_dec_35  \
count                 756.000000                 756.000000   
mean                   14.799230                  14.751559   
std                    15.722502                  14.432979   
min                     1.861700                   1.955900   
25%                     3.665925                   3.741275   
50%                     6.725700                   7.334250   
75%                    21.922050                  22.495175   
max                    62.007300                  57.544300   

       tqwt_kurtosisValue_dec_36       class  
count                 756.000000  756.000000  
mean                   31.481110    0.746032  
std                    34.230991    0.435568  
min                     2.364000    0.000000  
25%                  

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 756 entries, 0 to 755
Columns: 755 entries, id to class
dtypes: float64(749), int64(6)
memory usage: 4.4 MB


In [33]:
few_values = {}

for col in df:
    few_values[col] = pd.Series(df[col].value_counts().index.values)
        
few_values_df = pd.DataFrame(few_values).fillna('')
few_values_df.T

0        1        2        3        4    \
id                             0.0    173.0    160.0    161.0    162.0   
gender                         1.0      0.0                              
PPE                        0.82273  0.79067   0.7003  0.83918  0.79132   
DFA                        0.72248  0.80358  0.80616  0.75399  0.75192   
RPDE                       0.62128  0.45816  0.34552  0.42443  0.42171   
...                            ...      ...      ...      ...      ...   
tqwt_kurtosisValue_dec_33   3.1144   3.1761   3.1704   3.6458  30.2127   
tqwt_kurtosisValue_dec_34   12.595   3.1854   6.2712   50.159   4.9353   
tqwt_kurtosisValue_dec_35   4.2391   3.7708   4.6983   3.0004  10.4367   
tqwt_kurtosisValue_dec_36   3.2941  10.0693  87.2839  65.5773  91.8193   
class                          1.0      0.0                              

                               5        6        7        8        9    ...  \
id                           163.0    164.0    165.0    166.0    167.0  ...   
gender                                                                  ...   
PPE                        0.81783   0.8088   0.7993  0.79855  0.84892  ...   
DFA                        0.67772  0.67512  0.76831  0.67496  0.65177  ...   
RPDE                       0.70689  0.38853   0.6425  0.37645  0.34446  ...   
...                            ...      ...      ...      ...      ...  ...   
tqwt_kurtosisValue_dec_33   2.5687  61.8121  67.4397   3.2354  64.7489  ...   
tqwt_kurtosisValue_dec_34  19.5462  55.5402  56.0181   3.4099  51.1687  ...   
tqwt_kurtosisValue_dec_35  11.4409  28.0077  47.0509  46.4851   2.9277  ...   
tqwt_kurtosisValue_dec_36  93.9577   8.1257   95.085  92.9395  12.5873  ...   
class                                                                   ...   

                               745      746      747      748      749  \
id                                                                       
gender                                                                   
PPE                                                                      
DFA                                                                      
RPDE                       0.39573  0.31554  0.46374                     
...                            ...      ...      ...      ...      ...   
tqwt_kurtosisValue_dec_33  24.8386   9.7772  11.0156   4.1248   3.6401   
tqwt_kurtosisValue_dec_34   17.743  25.0281   11.091  23.3673   7.2245   
tqwt_kurtosisValue_dec_35  19.0558  24.7547   8.1797  23.4082   6.6474   
tqwt_kurtosisValue_dec_36  64.1884  69.1076  68.6664  46.1423  10.0114   
class                                                                    

                              750     751      752     753 754  
id                                                              
gender                                                          
PPE                                                             
DFA                                                             
RPDE                                                            
...                           ...     ...      ...     ...  ..  
tqwt_kurtosisValue_dec_33  2.9474  3.0415                       
tqwt_kurtosisValue_dec_34  8.6534  4.3563   4.0116              
tqwt_kurtosisValue_dec_35  4.2132  3.5039   2.6217              
tqwt_kurtosisValue_dec_36  6.9022  3.5402  28.7122  3.1527      
class                                                           

[755 rows x 755 columns]

### Output:
Dataset has dimensions (756, 755). 756 rows, 755 attributes. All attributes are numeric.

## 4. Check the set for missing, empty values ​​(command "isnull().sum()". See which attributes have the correct data type and which ones need to be changed.

In [37]:
df.isnull().sum()

id                           0
gender                       0
PPE                          0
DFA                          0
RPDE                         0
                            ..
tqwt_kurtosisValue_dec_33    0
tqwt_kurtosisValue_dec_34    0
tqwt_kurtosisValue_dec_35    0
tqwt_kurtosisValue_dec_36    0
class                        0
Length: 755, dtype: int64

In [39]:
df_describe.head(10)

id      gender         PPE         DFA        RPDE   numPulses  \
count  756.000000  756.000000  756.000000  756.000000  756.000000  756.000000   
mean   125.500000    0.515873    0.746284    0.700414    0.489058  323.972222   
std     72.793721    0.500079    0.169294    0.069718    0.137442   99.219059   
min      0.000000    0.000000    0.041551    0.543500    0.154300    2.000000   
25%     62.750000    0.000000    0.762833    0.647053    0.386537  251.000000   
50%    125.500000    1.000000    0.809655    0.700525    0.484355  317.000000   
75%    188.250000    1.000000    0.834315    0.754985    0.586515  384.250000   
max    251.000000    1.000000    0.907660    0.852640    0.871230  907.000000   

       numPeriodsPulses  meanPeriodPulses  stdDevPeriodPulses  locPctJitter  \
count        756.000000        756.000000          756.000000    756.000000   
mean         322.678571          0.006360            0.000383      0.002324   
std           99.402499          0.001826            0.000728      0.002628   
min            1.000000          0.002107            0.000011      0.000210   
25%          250.000000          0.005003            0.000049      0.000970   
50%          316.000000          0.006048            0.000077      0.001495   
75%          383.250000          0.007528            0.000171      0.002520   
max          905.000000          0.012966            0.003483      0.027750   

       ...  tqwt_kurtosisValue_dec_28  tqwt_kurtosisValue_dec_29  \
count  ...                 756.000000                 756.000000   
mean   ...                  26.237251                  22.840337   
std    ...                  42.220693                  32.626464   
min    ...                   1.509800                   1.531700   
25%    ...                   2.408675                   3.452800   
50%    ...                   5.586300                   7.062750   
75%    ...                  28.958075                  29.830850   
max    ...                 239.788800                 203.311300   

       tqwt_kurtosisValue_dec_30  tqwt_kurtosisValue_dec_31  \
count                 756.000000                 756.000000   
mean                   18.587888                  13.872018   
std                    25.537464                  20.046029   
min                     1.582900                   1.747200   
25%                     3.354825                   3.077450   
50%                     6.077400                   4.770850   
75%                    21.944050                  13.188000   
max                   121.542900                 102.207000   

       tqwt_kurtosisValue_dec_32  tqwt_kurtosisValue_dec_33  \
count                 756.000000                 756.000000   
mean                   12.218953                  12.375335   
std                    17.783642                  16.341665   
min                     1.789500                   1.628700   
25%                     2.937025                   3.114375   
50%                     4.300450                   4.741450   
75%                    10.876150                  12.201325   
max                    85.571700                  73.532200   

       tqwt_kurtosisValue_dec_34  tqwt_kurtosisValue_dec_35  \
count                 756.000000                 756.000000   
mean                   14.799230                  14.751559   
std                    15.722502                  14.432979   
min                     1.861700                   1.955900   
25%                     3.665925                   3.741275   
50%                     6.725700                   7.334250   
75%                    21.922050                  22.495175   
max                    62.007300                  57.544300   

       tqwt_kurtosisValue_dec_36       class  
count                 756.000000  756.000000  
mean                   31.481110    0.746032  
std                    34.230991    0.435568  
min                     2.364000    0.000000  
25%                  

### Output:
There are no missing attributes. All attributes are numeric.

## 6. Split the data from "x" and "y" into a training and test sample. To do this, use the "train_test_split" function. It takes the two previously specified variables x and y as arguments, and here we also define the size of the test sample (usually no more than 20%). To do this, we assign our value in the format "0.20" (which means 20%) to the named argument "test_size". In the "random_state" parameter, for the sake of consistency, we set a random initial number for the random number generator.

In [43]:
X = df.drop(columns=['id', 'class'])
y = df['class']

In [45]:
X.head()

gender      PPE      DFA     RPDE  numPulses  numPeriodsPulses  \
0       1  0.85247  0.71826  0.57227        240               239   
1       1  0.76686  0.69481  0.53966        234               233   
2       1  0.85083  0.67604  0.58982        232               231   
3       0  0.41121  0.79672  0.59257        178               177   
4       0  0.32790  0.79782  0.53028        236               235   

   meanPeriodPulses  stdDevPeriodPulses  locPctJitter  locAbsJitter  ...  \
0          0.008064            0.000087       0.00218      0.000018  ...   
1          0.008258            0.000073       0.00195      0.000016  ...   
2          0.008340            0.000060       0.00176      0.000015  ...   
3          0.010858            0.000183       0.00419      0.000046  ...   
4          0.008162            0.002669       0.00535      0.000044  ...   

   tqwt_kurtosisValue_dec_27  tqwt_kurtosisValue_dec_28  \
0                     1.5466                     1.5620   
1                     1.5530                     1.5589   
2                     1.5399                     1.5643   
3                     6.9761                     3.7805   
4                     7.8832                     6.1727   

   tqwt_kurtosisValue_dec_29  tqwt_kurtosisValue_dec_30  \
0                     2.6445                     3.8686   
1                     3.6107                    23.5155   
2                     2.3308                     9.4959   
3                     3.5664                     5.2558   
4                     5.8416                     6.0805   

   tqwt_kurtosisValue_dec_31  tqwt_kurtosisValue_dec_32  \
0                     4.2105                     5.1221   
1                    14.1962                    11.0261   
2                    10.7458                    11.0177   
3                    14.0403                     4.2235   
4                     5.7621                     7.7817   

   tqwt_kurtosisValue_dec_33  tqwt_kurtosisValue_dec_34  \
0                     4.4625                     2.6202   
1                     9.5082                     6.5245   
2                     4.8066                     2.9199   
3                     4.6857                     4.8460   
4                    11.6891                     8.2103   

   tqwt_kurtosisValue_dec_35  tqwt_kurtosisValue_dec_36  
0                     3.0004                    18.9405  
1                     6.3431                    45.1780  
2                     3.1495                     4.7666  
3                     6.2650                     4.0603  
4                     5.0559                     6.1164  

[5 rows x 753 columns]

In [47]:
y.head()

0    1
1    1
2    1
3    1
4    1
Name: class, dtype: int64

In [49]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

### Conclusion:
Training and testing kits have been prepared

### 7. After preparing the data set, design the neural network model. To do this, create a class named "Perceptron". For this class, define the "build" method with two parameters:
1) dim (input size of the training sample),
2) num_classes - the number of classes (in our case, there are 3) Next, in the body of this method, create a variable model, where we save an instance of the Sequential class from the Keras-Tensorflow library. Then, using the "add" method, we build the fully connected layers (Dense) of the neural network one by one. The first layer (input layer) will have 128 nodes (neurons) and the ReLU activation function. This layer also accepts the above-mentioned dim argument as a parameter. The second (hidden) layer has 32 neurons and the ReLU activation function. The third layer that completes the network has 2 neurons (according to the number of classes) and the Softmax activation function.

In [55]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Input, Dense, Activation, Dropout
from tensorflow.keras.optimizers import Adam


2025-03-27 18:50:08.424025: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:477] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1743090608.501761   10988 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1743090608.522884   10988 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2025-03-27 18:50:08.691003: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [57]:

class Perceptron:
    @staticmethod
    def build (dim, num_classes):
        model = Sequential()
        model.add(Dense(128, input_dim=dim, activation='relu'))
        model.add(Dense(32, activation='relu'))
        model.add(Dense(num_classes, activation='softmax'))
        
        return model

## 8. Now create an instance of the Perceptron class, storing it in a variable named model (you can use any variable name, but it is recommended to specify names that will most clearly describe their content). Don't forget to define the parameters of the Perceptron class: num_classes should match the number of classes in the dataset, and for dim we pass the input sample size x_train.shape[1], which is 753.

In [60]:
print(X_train.shape)

(604, 753)


In [62]:
model = Perceptron.build(X_train.shape[1],2)

/home/nikita/anaconda3/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
2025-03-27 18:50:42.158382: E external/local_xla/xla/stream_executor/cuda/cuda_driver.cc:152] failed call to cuInit: INTERNAL: CUDA error: Failed call to cuInit: UNKNOWN ERROR (303)


### Output
Created an instance of the Perceptron class (our model)

## 9. Compile the newly created neural network model. To do this, use the “compile” method (Keras-Tensorflow). As its first parameter, specify which error function to use (for example, “sparse_categorical_crossentropy”, which calculates the crossentropy loss between labels and predictions). As the second parameter, specify the optimizer (for example, “adam” or stochastic gradient descent). For information on loss functions and optimizers, see the Keras, Tensorflow documentation. And as the third parameter, specify the target metric – accuracy.

In [66]:
model.compile(loss='sparse_categorical_crossentropy', optimizer=Adam(0.00001), metrics=['accuracy'])
print("Model compiled")

Model compiled


### Output:
Model compiled.

## 10. Now, after completing all the preparatory operations, you can begin the process of training the ANN model. To do this, apply the fit method (Keras, Tensorflow) to the model variable and specify the applied samples (x_train, y_train), batch size, number of iterations (epochs), size of the validation sample "validation_split" (which is created automatically by this fit method) as parameters, and also define the verbose parameter (by setting it to 0, 1 or 2, we choose how we want to "see" the training progress for each epoch).

In [71]:
r = model.fit(X_train, y_train, batch_size=32, epochs=200, verbose=1,validation_data=(X_test,y_test))

Epoch 1/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 1s 7ms/step - accuracy: 0.7637 - loss: 148349856.0000 - val_accuracy: 0.7500 - val_loss: 149205872.0000
Epoch 2/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7521 - loss: 127366272.0000 - val_accuracy: 0.7500 - val_loss: 125259048.0000
Epoch 3/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7594 - loss: 109507320.0000 - val_accuracy: 0.7500 - val_loss: 99714248.0000
Epoch 4/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7309 - loss: 88136888.0000 - val_accuracy: 0.7500 - val_loss: 74637088.0000
Epoch 5/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7326 - loss: 70035312.0000 - val_accuracy: 0.7500 - val_loss: 48667324.0000
Epoch 6/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.7550 - loss: 38196588.0000 - val_accuracy: 0.7500 - val_loss: 24403840.0000
Epoch 7/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7475 - loss: 17249766.0000 - val_accuracy: 0.7303 - val_loss: 810758.7500
Epoch 8/20

19/19 [==============================] - 0s 7ms/step - loss: 302107.4062 - accuracy: 0.6275 - val_loss: 306956.2188 - val_accuracy: 0.6579
Epoch 163/200
19/19 [==============================] - 0s 7ms/step - loss: 307886.7188 - accuracy: 0.6109 - val_loss: 313290.4062 - val_accuracy: 0.5592
Epoch 164/200
19/19 [==============================] - 0s 12ms/step - loss: 353653.6562 - accuracy: 0.6175 - val_loss: 314702.0938 - val_accuracy: 0.6776
Epoch 165/200
19/19 [==============================] - 0s 8ms/step - loss: 371540.5312 - accuracy: 0.6043 - val_loss: 307070.0938 - val_accuracy: 0.5461
Epoch 166/200
19/19 [==============================] - 0s 13ms/step - loss: 321064.4375 - accuracy: 0.5712 - val_loss: 322264.0938 - val_accuracy: 0.7237
Epoch 167/200
19/19 [==============================] - 0s 8ms/step - loss: 310680.4062 - accuracy: 0.6325 - val_loss: 316313.5938 - val_accuracy: 0.7105
Epoch 168/200
19/19 [==============================] - 0s 13ms/step - loss: 307024.6875 - accu

### Conclusion:
The model training process has been completed

## 11. After the training process is complete, evaluate its quality. To do this, apply the "evaluate" method (Keras, Tensorflow) to the model.

In [75]:
pd.DataFrame(r.history)

accuracy          loss  val_accuracy      val_loss
0    0.745033  1.507114e+08      0.750000  1.492059e+08
1    0.745033  1.283586e+08      0.750000  1.252590e+08
2    0.745033  1.055207e+08      0.750000  9.971425e+07
3    0.745033  8.181851e+07      0.750000  7.463709e+07
4    0.745033  5.908621e+07      0.750000  4.866732e+07
..        ...           ...           ...           ...
195  0.682119  4.731178e+05      0.723684  4.982391e+05
196  0.653974  4.795352e+05      0.717105  4.395675e+05
197  0.720199  3.967176e+05      0.684211  3.665788e+05
198  0.713576  3.743529e+05      0.664474  3.726475e+05
199  0.735099  3.659940e+05      0.644737  4.091752e+05

[200 rows x 4 columns]

In [77]:
import hvplot.pandas
pd.DataFrame(r.history).hvplot.line(y=['loss', 'val_loss'])

:NdOverlay   [Variable]
   :Curve   [index]   (value)

In [79]:
print('\n# We evaluate on test data')
results = model.evaluate(X_test, y_test, batch_size=128)
print('test loss, test acc:', results)


# We evaluate on test data
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - accuracy: 0.6512 - loss: 401193.6875
test loss, test acc: [409176.4375, 0.6447368264198303]


In [81]:
def print_evaluate(true, predicted):  
    mae = metrics.mean_absolute_error(true, predicted)
    mse = metrics.mean_squared_error(true, predicted)
    rmse = np.sqrt(metrics.mean_squared_error(true, predicted))
    r2_square = metrics.r2_score(true, predicted)
    print('MAE:', mae)
    print('MSE:', mse)
    print('RMSE:', rmse)
    print('R2 Square', r2_square)
    print('__________________________________')
    
def evaluate(true, predicted):
    mae = metrics.mean_absolute_error(true, predicted)
    mse = metrics.mean_squared_error(true, predicted)
    rmse = np.sqrt(metrics.mean_squared_error(true, predicted))
    r2_square = metrics.r2_score(true, predicted)
    return mae, mse, rmse, r2_square

In [83]:
test_pred = model.predict(X_test)
train_pred = model.predict(X_train)


df_temp1 = pd.DataFrame(test_pred, columns=['1', '2'])
df_temp2 = pd.DataFrame(train_pred, columns=['1', '2'])

print('Test set evaluation:\n_____________________________________')
print_evaluate(y_test, df_temp1['2'])

print('Train set evaluation:\n_____________________________________')
print_evaluate(y_train, df_temp2['2'])

results_df = pd.DataFrame(data=[["Artficial Neural Network", *evaluate(y_test, df_temp1['2']), 0]], 
                            columns=['Model', 'MAE', 'MSE', 'RMSE', 'R2 Square', 'Cross Validation'])
results_df.head()

5/5 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step 
Test set evaluation:
_____________________________________
MAE: 0.3552631814228861
MSE: 0.35526314064076314
RMSE: 0.5960395462054201
R2 Square -0.8947367500840702
__________________________________
Train set evaluation:
_____________________________________
MAE: 0.27152320693265525
MSE: 0.2715231602555107
RMSE: 0.5210788426481262
R2 Square -0.429379411713916
__________________________________


Model       MAE       MSE     RMSE  R2 Square  \
0  Artficial Neural Network  0.355263  0.355263  0.59604  -0.894737   

   Cross Validation  
0                 0

## 12. Use the matplotlib library to plot the training and testing graphs of the model.

In [86]:
import matplotlib as plt

In [88]:
import hvplot.pandas
pd.DataFrame(r.history).hvplot.line(y=['accuracy', 'val_accuracy'])

:NdOverlay   [Variable]
   :Curve   [index]   (value)

## 13. Try standardizing the training and test samples. To do this, use the StandardScaler function from the sklearn library. For comparison, apply the standardized sample to the ANN model, retrain it and plot the graphs, evaluate the quality metrics, compare them with the previous results, whether the accuracy has improved after standardizing the data set.

In [91]:
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
pipeline = Pipeline([
    ('std_scalar', StandardScaler())
])

X_train = pipeline.fit_transform(X_train)
X_test = pipeline.transform(X_test)

In [93]:
model = Perceptron.build(X_train.shape[1],2)

/home/nikita/anaconda3/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [95]:
model.compile(loss='sparse_categorical_crossentropy', optimizer=Adam(0.00001), metrics=['accuracy'])
print("Model compiled")

Model compiled


In [97]:
r = model.fit(X_train, y_train, batch_size=32, epochs=200, verbose=1,validation_data=(X_test,y_test))

Epoch 1/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 1s 7ms/step - accuracy: 0.5498 - loss: 0.7196 - val_accuracy: 0.6316 - val_loss: 0.6452
Epoch 2/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6165 - loss: 0.6866 - val_accuracy: 0.6513 - val_loss: 0.6241
Epoch 3/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6012 - loss: 0.6724 - val_accuracy: 0.6513 - val_loss: 0.6064
Epoch 4/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6136 - loss: 0.6498 - val_accuracy: 0.6711 - val_loss: 0.5908
Epoch 5/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6282 - loss: 0.6353 - val_accuracy: 0.6842 - val_loss: 0.5763
Epoch 6/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.6802 - loss: 0.6062 - val_accuracy: 0.6908 - val_loss: 0.5644
Epoch 7/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7042 - loss: 0.5836 - val_accuracy: 0.7105 - val_loss: 0.5532
Epoch 8/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6845 - loss: 0.5865 - val_accuracy: 0.7171 - v

19/19 [==============================] - 0s 10ms/step - loss: 0.0785 - accuracy: 0.9983 - val_loss: 0.3603 - val_accuracy: 0.8487
Epoch 173/200
19/19 [==============================] - 0s 6ms/step - loss: 0.0777 - accuracy: 0.9983 - val_loss: 0.3607 - val_accuracy: 0.8487
Epoch 174/200
19/19 [==============================] - 0s 9ms/step - loss: 0.0769 - accuracy: 0.9983 - val_loss: 0.3606 - val_accuracy: 0.8487
Epoch 175/200
19/19 [==============================] - 0s 8ms/step - loss: 0.0761 - accuracy: 0.9983 - val_loss: 0.3605 - val_accuracy: 0.8487
Epoch 176/200
19/19 [==============================] - 0s 10ms/step - loss: 0.0753 - accuracy: 0.9983 - val_loss: 0.3600 - val_accuracy: 0.8487
Epoch 177/200
19/19 [==============================] - 0s 7ms/step - loss: 0.0745 - accuracy: 0.9983 - val_loss: 0.3607 - val_accuracy: 0.8487
Epoch 178/200
19/19 [==============================] - 0s 11ms/step - loss: 0.0738 - accuracy: 0.9983 - val_loss: 0.3604 - val_accuracy: 0.8487
Epoch 179/

In [98]:
pd.DataFrame(r.history)

accuracy      loss  val_accuracy  val_loss
0    0.566225  0.719788      0.631579  0.645221
1    0.600993  0.689991      0.651316  0.624082
2    0.612583  0.663218      0.651316  0.606421
3    0.639073  0.639615      0.671053  0.590765
4    0.658940  0.618170      0.684211  0.576339
..        ...       ...           ...       ...
195  0.993378  0.089118      0.861842  0.336981
196  0.993378  0.088278      0.861842  0.336517
197  0.993378  0.087488      0.861842  0.336490
198  0.993378  0.086705      0.861842  0.336206
199  0.995033  0.085974      0.861842  0.336364

[200 rows x 4 columns]

In [99]:
import hvplot.pandas
pd.DataFrame(r.history).hvplot.line(y=['loss', 'val_loss'])

:NdOverlay   [Variable]
   :Curve   [index]   (value)

In [103]:
print('\n# We evaluate on test data')
results = model.evaluate(X_test, y_test, batch_size=128)
print('test loss, test acc:', results)


# We evaluate on test data
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.8610 - loss: 0.3395
test loss, test acc: [0.3363634943962097, 0.8618420958518982]


In [105]:
import hvplot.pandas
pd.DataFrame(r.history).hvplot.line(y=['accuracy', 'val_accuracy'])

:NdOverlay   [Variable]
   :Curve   [index]   (value)

## 14. Try training the model with different parameters: change the error function, optimizer (choose from those presented in the Keras documentation), the number of epochs (iterations). Also try changing the number of neurons in the input and hidden layers. Analyze and compare the results, plot graphs.

In [108]:

class Perceptron2:
    @staticmethod
    def build (dim, num_classes):
        model = Sequential()
        model.add(Dense(128, input_dim=dim, activation='relu'))
        model.add(Dense(64, activation='relu'))
        model.add(Dense(32, activation='relu'))
        model.add(Dense(num_classes, activation='softmax'))
        
        return model

In [110]:
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
pipeline = Pipeline([
    ('std_scalar', StandardScaler())
])

X_train = pipeline.fit_transform(X_train)
X_test = pipeline.transform(X_test)

In [112]:
model = Perceptron2.build(X_train.shape[1],2)

/home/nikita/anaconda3/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [114]:
from tensorflow import keras

In [116]:
model.compile(loss=keras.losses.SparseCategoricalCrossentropy(),
              optimizer=keras.optimizers.RMSprop(learning_rate=1e-3), 
             metrics=[keras.metrics.SparseCategoricalAccuracy()])
print("Model compiled")

Model compiled


In [118]:
r = model.fit(X_train, y_train, batch_size=32, epochs=200, verbose=1,validation_data=(X_test,y_test))

Epoch 1/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 1s 12ms/step - loss: 0.5284 - sparse_categorical_accuracy: 0.7380 - val_loss: 0.3471 - val_sparse_categorical_accuracy: 0.8684
Epoch 2/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 0.2073 - sparse_categorical_accuracy: 0.9281 - val_loss: 0.3516 - val_sparse_categorical_accuracy: 0.8684
Epoch 3/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 0.1273 - sparse_categorical_accuracy: 0.9642 - val_loss: 0.2979 - val_sparse_categorical_accuracy: 0.8816
Epoch 4/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 0.0442 - sparse_categorical_accuracy: 0.9973 - val_loss: 0.3934 - val_sparse_categorical_accuracy: 0.8750
Epoch 5/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 0.0536 - sparse_categorical_accuracy: 0.9802 - val_loss: 0.4069 - val_sparse_categorical_accuracy: 0.8553
Epoch 6/200
19/19 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 0.0074 - sparse_categorical_accuracy: 1.0000 - val_loss: 0.4349 - val_sparse_categorical_accuracy: 0.8947
Epoch 7/2

19/19 [==============================] - 0s 7ms/step - loss: 0.0000e+00 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.5663 - val_sparse_categorical_accuracy: 0.8882
Epoch 91/200
19/19 [==============================] - 0s 12ms/step - loss: 0.0000e+00 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.5675 - val_sparse_categorical_accuracy: 0.8882
Epoch 92/200
19/19 [==============================] - 0s 8ms/step - loss: 0.0000e+00 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.5687 - val_sparse_categorical_accuracy: 0.8882
Epoch 93/200
19/19 [==============================] - 0s 10ms/step - loss: 0.0000e+00 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.5698 - val_sparse_categorical_accuracy: 0.8882
Epoch 94/200
19/19 [==============================] - 0s 8ms/step - loss: 0.0000e+00 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.5711 - val_sparse_categorical_accuracy: 0.8882
Epoch 95/200
19/19 [==============================] - 0s 9ms/step - loss: 0.0000e+00 - spar

19/19 [==============================] - 0s 11ms/step - loss: 0.0000e+00 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.6083 - val_sparse_categorical_accuracy: 0.8882
Epoch 135/200
19/19 [==============================] - 0s 9ms/step - loss: 0.0000e+00 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.6090 - val_sparse_categorical_accuracy: 0.8882
Epoch 136/200
19/19 [==============================] - 0s 11ms/step - loss: 0.0000e+00 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.6098 - val_sparse_categorical_accuracy: 0.8882
Epoch 137/200
19/19 [==============================] - 0s 13ms/step - loss: 0.0000e+00 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.6105 - val_sparse_categorical_accuracy: 0.8882
Epoch 138/200
19/19 [==============================] - 0s 9ms/step - loss: 0.0000e+00 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.6113 - val_sparse_categorical_accuracy: 0.8882
Epoch 139/200
19/19 [==============================] - 0s 13ms/step - loss: 0.0000e+00

19/19 [==============================] - 0s 11ms/step - loss: 0.0000e+00 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.6363 - val_sparse_categorical_accuracy: 0.8882
Epoch 179/200
19/19 [==============================] - 0s 13ms/step - loss: 0.0000e+00 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.6369 - val_sparse_categorical_accuracy: 0.8882
Epoch 180/200
19/19 [==============================] - 0s 7ms/step - loss: 0.0000e+00 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.6374 - val_sparse_categorical_accuracy: 0.8882
Epoch 181/200
19/19 [==============================] - 0s 10ms/step - loss: 0.0000e+00 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.6379 - val_sparse_categorical_accuracy: 0.8882
Epoch 182/200
19/19 [==============================] - 0s 9ms/step - loss: 0.0000e+00 - sparse_categorical_accuracy: 1.0000 - val_loss: 1.6385 - val_sparse_categorical_accuracy: 0.8882
Epoch 183/200
19/19 [==============================] - 0s 9ms/step - loss: 0.0000e+00 

In [119]:
pd.DataFrame(r.history)

loss  sparse_categorical_accuracy  val_loss  \
0    4.555498e-01                     0.794702  0.347064   
1    2.110623e-01                     0.922185  0.351640   
2    1.137191e-01                     0.965232  0.297929   
3    5.048985e-02                     0.991722  0.393408   
4    3.320033e-02                     0.991722  0.406939   
..            ...                          ...       ...   
195  6.797284e-07                     1.000000  0.883013   
196  6.745969e-07                     1.000000  0.883349   
197  6.708470e-07                     1.000000  0.883713   
198  6.667024e-07                     1.000000  0.884090   
199  6.635445e-07                     1.000000  0.884456   

     val_sparse_categorical_accuracy  
0                           0.868421  
1                           0.868421  
2                           0.881579  
3                           0.875000  
4                           0.855263  
..                               ...  
195                         0.888158  
196                         0.888158  
197                         0.888158  
198                         0.888158  
199                         0.888158  

[200 rows x 4 columns]

In [120]:
import hvplot.pandas
pd.DataFrame(r.history).hvplot.line(y=['loss', 'val_loss'])

:NdOverlay   [Variable]
   :Curve   [index]   (value)

In [124]:
print('\n# We evaluate on test data')
results = model.evaluate(X_test, y_test, batch_size=128)
print('test loss, test acc:', results)


# We evaluate on test data
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - loss: 0.9098 - sparse_categorical_accuracy: 0.8838
test loss, test acc: [0.8844559192657471, 0.8881579041481018]


### Output:
The model with an additional layer has been launched

## 15. Draw a final conclusion – with what parameters did the best metrics of the quality of ANN training turn out and did standardization of the data set help.

### Conclusion.
Standardization of the model increases the convergence rate, as a result, fewer epochs were needed to achieve an acceptable result.